filebeat/backup/registry.go

```diff
@@ -0,0 +1,138 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package backup
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"sort"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+const (
+	regLogFilename = "log.json"
+)
+
+var (
+	errCheckpointNotFound = errors.New("there is no checkpoint in the registry")
+	checkpointFileRegexp  = regexp.MustCompile(`^[0-9]+\.json$`)
+)
+
+// NewRegistryBackuper creates a new backuper that creates backups for the registry files.
+// It creates backups for the checkpoint file (if exists) and the registry log.
+// `regHome` must be the final directory with the registry log (`log.json`).
+func NewRegistryBackuper(log *logp.Logger, regHome string) Backuper {
+	return &registryBackuper{
+		log:     log,
+		regHome: regHome,
+	}
+}
+
+type registryBackuper struct {
+	log             *logp.Logger
+	regHome         string
+	removeCallbacks []func() error
+}
+
+// Backup backs up the active checkpoint if any and the registry log file
+func (rb *registryBackuper) Backup() error {
+	var toBackup []string
+
+	rb.log.Debug("Attempting to find the checkpoint...")
+	checkpoint, err := rb.findCheckpoint()
+	if err == nil {
+		toBackup = append(toBackup, checkpoint)
+		rb.log.Debugf("Found checkpoint at %s", checkpoint)
+	} else if err != nil && !errors.Is(err, errCheckpointNotFound) {
+		return fmt.Errorf("failed to look for a checkpoint file in %s: %w", rb.regHome, err)
+	} else {
+		rb.log.Debug("Checkpoint not found")
+	}
+
+	registryLog := filepath.Join(rb.regHome, regLogFilename)
+	rb.log.Debugf("Checking if the registry log exists at %s...", registryLog)
+	exists, err := fileExists(registryLog)
```

Comment:
I think there is an interesting error conditions here.  what if the registryLog file is deleted after the call to fileExists, but before the Backup() is performed?  I think you will stop processing the Backup and error because the file is gone and the open will fail.  Is that the right behavior or should continue the backup?  Or should we pass an open filehandle so it won't matter if a delete happens?

Reply:
That's an interesting point, but I believe the assumption is that nothing else is working on those files while the backup/registry migration is happening. I haven't gone through the whole PR but I believe that is a fair enough assumption and that should be enforced when creating the backups migrating the registry.

I believe we do not need to worry about external intervention on those files.

Anyway, let's wait for @rdner thoughts on this ;)

Reply:
While it's a generic backup implementation and theoretically someone could mis-use it, it's not the case with this PR. 

If we're talking about this PR, the code is executed before any input is running and the only case when a file can disappear is replacing an existent checkpoint which requires an input to run and is a rare event.

To summarise, I don't think there is a concern here and failing a backup will prevent any changes, so it's safe to fail during the backup and I think it's the right behaviour.